include/operations/CompoundOperation.hpp

```diff
@@ -54,6 +54,36 @@ class CompoundOperation final : public Operation {
     });
   }
 
+  void addControls(const Controls& c) override {
+    for (auto ctrl : c) {
+      controls.insert(ctrl);
+    }
+    // we can just add the controls to each operation, as the operations will
+    // check if they already act on the control qubits.
+    for (auto& op : ops) {
+      op->addControls(c);
+    }
+  }
+
+  void clearControls() override {
+    // we remove just our controls from nested operations
+    removeControls(Controls{controls});
```

Comment:
```suggestion
    removeControls(controls);
```
That should work as well, right? No need to construct a new object here.

Reply:
This is needed as otherwise the iterator in `removeControls` will be modified while we iterate through it.

Reply:
I should probably add a comment in there why I'm doing this :)

Reply:
Ah, that makes sense.
Although it is a only a very small cost, you could technically avoid that cost by provifing an overload for `removeControl` that takes an iterator to a control and returns an iterator.
That way you could replicate the behavior of `erase`, which returns a valid iterator to the element after the last removed element (see https://en.cppreference.com/w/cpp/container/set/erase).
Overall, iterator validity is an important concern though. This also directly relates to my suggestion of making the `removeControl` function abstract and simply calling it repeatedly in the `removeControls` function.